boringtun/src/noise/timers.rs

```diff
@@ -273,15 +275,20 @@ impl TunnInner {
                     }
                 }
 
-                // If we have sent a packet to a given peer but have not received a
+                // If we have sent a data packet to a given peer but have not received a
                 // packet after from that peer for (KEEPALIVE + REKEY_TIMEOUT) ms,
                 // we initiate a new handshake.
-                if data_packet_sent > aut_packet_received
-                    && now - aut_packet_received >= KEEPALIVE_TIMEOUT + REKEY_TIMEOUT
-                    && mem::replace(&mut self.timers.want_handshake, false)
+                if self
+                    .timers
+                    .want_handshake_since
+                    .map(|want_handshake_since| {
+                        (now - want_handshake_since) >= (KEEPALIVE_TIMEOUT + REKEY_TIMEOUT)
+                    })
+                    .unwrap_or_default()
                 {
                     tracing::warn!("HANDSHAKE(KEEPALIVE + REKEY_TIMEOUT)");
                     handshake_initiation_required = true;
+                    self.timers.want_handshake_since = None;
```

Comment:
If I understand this correctly, this would mean, that in this situation:
1. Send data packet
2. No data or keepalive packet received within KEEPALIVE_TIMEOUT + REKEY_TIMEOUT
3. Handshake initiation packet is created and sent
4. No response...
5. No handshake retry

That the handshake initiation is attempted only once instead of retrying until REKEY-ATTEMPT-TIME expires.

You can find more info here: https://www.wireguard.com/papers/wireguard.pdf#subsection.6.5

>  if a transport data message has not been received for (Keepalive-Timeout +
Rekey-Timeout) seconds, in which case a handshake initiation message is sent to the unresponsive peer,
once every Rekey-Timeout seconds, as in section 6.4, until a secure session is recreated successfully or until
Rekey-Attempt-Time seconds have passed.

Reply:
I think handshakes will be retried every `Rekey-Timeout` - they are retried (see [here](https://github.com/NordSecurity/boringtun/blob/35a3fb944eb6ad857ec3436db2f4a7a2eb827620/boringtun/src/noise/timers.rs#L238-L245)) until `handshake_initiation_required` is `true` and this PR doesn't change that.

Reply:
Could we devise a testcase to catch this?

Reply:
Yeah, @tomaszklak seems to be right, so testcase would not really catch this, as the situation is not there.